File: src/main/java/recipebase/server/recipe/model/Quantity.java
package recipebase.server.recipe.model;

import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.fasterxml.jackson.annotation.JsonTypeName;
import java.math.BigDecimal;

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
public sealed interface Quantity {
    @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
    sealed interface Unit {
        @JsonTypeName("gram")
        record Gram() implements Unit {}

        @JsonTypeName("kilogram")
        record Kilogram() implements Unit {}

        @JsonTypeName("liter")
        record Liter() implements Unit {}

        @JsonTypeName("milliliter")
        record Milliliter() implements Unit {}

        @JsonTypeName("arbitrary")
        record Arbitrary() implements Unit {}

        @JsonTypeName("custom")
        record Custom(String name) implements Unit {}
    }

    @JsonTypeName("decimal")
    record DecimalAmount(Unit unit, BigDecimal amount) implements Quantity {}

    @JsonTypeName("interval")
    record Interval(
        Unit unit,
        BigDecimal from,
        BigDecimal to
    ) implements Quantity {}

    @JsonTypeName("unspecified")
    record Unspecified(String notes) implements Quantity {}
}

File: frontend/src/app/shared/models/quantity.ts
/**
 * Type guards for generated Quantity types
 */
import type {
  Arbitrary,
  Custom,
  DecimalAmount,
  Gram,
  Interval,
  Kilogram,
  Liter,
  Milliliter,
  QuantityUnion as Quantity,
  UnitUnion as Unit,
  Unspecified,
} from "../server";

export type {
  Arbitrary,
  Custom,
  DecimalAmount,
  Gram,
  Interval,
  Kilogram,
  Liter,
  Milliliter,
  Quantity,
  Unit,
  Unspecified,
};

/**
 * Type guards for Quantity types
 */
export function isDecimalQuantity(q: Quantity): q is DecimalAmount {
  return q.type === "decimal";
}

export function isIntervalQuantity(q: Quantity): q is Interval {
  return q.type === "interval";
}

export function isUnspecifiedQuantity(q: Quantity): q is Unspecified {
  return q.type === "unspecified";
}

export function isStandardUnit(u: Unit): u is Gram | Kilogram | Liter | Milliliter {
  return u.type === "gram" || u.type === "kilogram" || u.type === "liter" || u.type === "milliliter";
}

export function isArbitraryUnit(u: Unit): u is Arbitrary {
  return u.type === "arbitrary";
}

export function isCustomUnit(u: Unit): u is Custom {
  return u.type === "custom";
}

/**
 * Get unit name for display
 */
export function getUnitName(unit: Unit): string {
  if (isCustomUnit(unit)) {
    return unit.name;
  }
  return unit.type;
}
